src/parser/jobs/sam/modules/Shoha.tsx

```diff
@@ -0,0 +1,272 @@
+import {t} from '@lingui/macro'
+import {Plural, Trans} from '@lingui/react'
+import Color from 'color'
+import React, {Fragment} from 'react'
+import {Accordion} from 'semantic-ui-react'
+
+import STATUSES from 'data/STATUSES'
+
+import {ActionLink} from 'components/ui/DbLink'
+import TimeLineChart from 'components/ui/TimeLineChart'
+import ACTIONS from 'data/ACTIONS'
+import JOBS from 'data/JOBS'
+import {CastEvent, Event} from 'fflogs'
+import Module, {dependency, DISPLAY_MODE} from 'parser/core/Module'
+import Checklist, {Requirement, Rule} from 'parser/core/modules/Checklist'
+import Combatants from 'parser/core/modules/Combatants'
+import {ComboEvent} from 'parser/core/modules/Combos'
+import Suggestions, {SEVERITY, TieredSuggestion} from 'parser/core/modules/Suggestions'
+
+const GENERATORS = {
+	[ACTIONS.HIGANBANA.id]: 1,
+	[ACTIONS.KAESHI_HIGANBANA.id]: 1,
+	[ACTIONS.TENKA_GOKEN.id]: 1,
+	[ACTIONS.KAESHI_GOKEN.id]: 1,
+	[ACTIONS.MIDARE_SETSUGEKKA.id]: 1,
+	[ACTIONS.KAESHI_SETSUGEKKA.id]: 1,
+}
+
+const SPENDERS = {
+	[ACTIONS.SHOHA.id]: 3,
+}
+
+const MAX_STACKS = 3
+
+class StackState {
```

Comment:
You're not actually declaring this as a class instance anymore, probably better as an `interface`. Alternatively make history a real class and have a method to pull data out of it so you get time data for wastage etc, and can still generate the charts nicely.